src/compression.rs

```diff
@@ -1189,13 +1180,63 @@ impl CompressionStepResult {
+    pub inv: Invention,
+    pub done: FinishedPattern,
+    pub dc_inv_str: String,
+    pub train_data: CompressionStepData,
+    pub test_data: Option<CompressionStepData>,
+}
+
+impl CompressionStepResult {
+    fn new(done: FinishedPattern, train_programs_node: Id, test_programs_node: Option<Id>, inv_name: &str, shared: &mut SharedData, past_invs: &Vec<CompressionStepResult>, first_train_cost: Option<i32>, first_test_cost: Option<i32>) -> Self {
+
+        let inv = done.to_invention(inv_name, shared);
+
+        let train_data = CompressionStepData::new(done.clone(), train_programs_node, inv_name, shared, past_invs, first_train_cost);
+        let test_data = test_programs_node.map(|n| CompressionStepData::new(done.clone(), n, inv_name, shared, past_invs, first_test_cost));
+
+        let dc_inv_str: String = dc_inv_str(&inv, past_invs);
+
+        CompressionStepResult { inv, done, dc_inv_str, train_data, test_data }
+    }
+
+    pub fn json(&self) -> serde_json::Value {        
+
+        json!({            
+            "body": self.inv.body.to_string(),
+            "dreamcoder": self.dc_inv_str,
+            "arity": self.inv.arity,
+            "name": self.inv.name,
+            "utility": self.done.utility,
+            "train_result": self.train_data.json(&self.inv.name),
+            "test_result": if let Some(d) = &self.test_data { d.json(&self.inv.name).to_string() } else { "N/A".to_string() },
+        })
+    }
+}
+
 impl fmt::Display for CompressionStepResult {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        if self.expected_cost != self.final_cost {
-            write!(f,"[cost mismatch of {}] ", self.expected_cost - self.final_cost)?;
+        if self.train_data.expected_cost != self.train_data.final_cost {
+            write!(f,"[cost mismatch of {} in training data] ", self.train_data.expected_cost - self.train_data.final_cost)?;
+        }
+        if let Some(test_d) = &self.test_data {
+            if test_d.expected_cost != test_d.final_cost {
+                write!(f,"[cost mismatch of {} in test data] ", test_d.expected_cost - test_d.final_cost);
+            }
         }
-        write!(f, "utility: {} | final_cost: {} | {:.2}x | uses: {} | body: {}",
-            self.done.utility, self.final_cost, self.multiplier, self.uses, self.inv)
+        write!(f, "utility: {} | final_cost (train): {} | multiplier (train): {:.2}x | uses (train): {} | final_cost(test): {} | multiplier (test): {:.2}x | uses (test): {} | body: {}",
+            self.done.utility,
+            self.train_data.final_cost,
+            self.train_data.multiplier,
+            self.train_data.uses,
+            if let Some(d) = &self.test_data { d.final_cost.to_string() } else { "N/A".to_string() },
+            if let Some(d) = &self.test_data { d.multiplier.to_string() } else { "N/A".to_string() },
+            if let Some(d) = &self.test_data { d.uses.to_string() } else { "N/A".to_string() },
```

Comment:
See comment for line 96 in `compress.rs`.

Reply:
`as_deref` and `map_or` with `null` for all these as described above